README.md
## ethjs-unit

<div>
  <!-- Build Status -->
  <a href="https://travis-ci.org/ethjs/ethjs-unit">
    <img src="https://travis-ci.org/ethjs/ethjs-unit.svg"
    alt="Build Status" />
  </a>

  <!-- NPM Version -->
  <a href="https://www.npmjs.org/package/ethjs-unit">
    <img src="http://img.shields.io/npm/v/@normalizex/ethjs-unit.svg"
    alt="NPM version" />
  </a>

  <!-- Test Coverage -->
  <a href="https://coveralls.io/r/ethjs/ethjs-unit">
    <img src="https://coveralls.io/repos/github/ethjs/ethjs-unit/badge.svg" alt="Test Coverage" />
  </a>

  <!-- Javascript Style -->
  <a href="http://airbnb.io/javascript/">
    <img src="https://img.shields.io/badge/code%20style-airbnb-brightgreen.svg" alt="js-airbnb-style" />
  </a>
</div>

<br />

A simple module for handling Ethereum unit convertion.

## Install

Using npm:
```console
npm install  @normalizex/ethjs-unit
```
Using yarn:
```console
yarn add @normalizex/ethjs-unit
```

## Usage

```js
const unit = require('@normalizex/ethjs-unit');

unit.toWei(249824778, 'ether');
// Result: '249824778000000000000000000'

unit.fromWei('249824778000000000000000000', 'ether');
// Result: '249824778'
```

##### For non-standard unit values ​​that are not in unitMap
```js
unit.toWeiByDecimals('24032190.064558267259', 12);
// Result '24032190064558267259'

unit.fromWeiByDecimals('24032190064558267259', 12);
// Result: '24032190.064558267259'

```

## About

A port from the `web3.js` library, that just handles the unit convertion between the various types of Ethereum currency units.

`toWei`, `fromWei`, `toWeiByDecimals`, `fromWeiByDecimals` allways return string

## Amorphic Data Formatting

`ethjs-unit` uses the [number-to-bn](http://github.com/silentcicero/number-to-bn) module to format all number values (hex or otherwise) into digestable BN.js number instances.

## Methods Available & Objects

```
unitMap           { unitName: singleUnitWeiValue, ... }
getValueOfUnit    <Function (unit) : (BN)>
toWei             <Function (value, unit) : (String)>
toWeiByDecimals   <Function (value, unit) : (String)>
fromWei           <Function (value, unit) : (String)>
fromWeiByDecimals <Function (value, unit) : (String)>

```

## Supported Units

```
'wei':          '1',
'kwei':         '1000',
'Kwei':         '1000',
'babbage':      '1000',
'femtoether':   '1000',
'mwei':         '1000000',
'Mwei':         '1000000',
'lovelace':     '1000000',
'picoether':    '1000000',
'gwei':         '1000000000',
'Gwei':         '1000000000',
'shannon':      '1000000000',
'nanoether':    '1000000000',
'nano':         '1000000000',
'szabo':        '1000000000000',
'microether':   '1000000000000',
'micro':        '1000000000000',
'finney':       '1000000000000000',
'milliether':   '1000000000000000',
'milli':        '1000000000000000',
'ether':        '1000000000000000000',
'kether':       '1000000000000000000000',
'grand':        '1000000000000000000000',
'mether':       '1000000000000000000000000',
'gether':       '1000000000000000000000000000',
'tether':       '1000000000000000000000000000000'
```

## Why BN.js?

`ethjs` has made a policy of using `BN.js` accross all of our modules. Here are some reasons why:

  1. Lighter than alternatives (BigNumber.js)
  2. Faster than most alternatives, see [benchmarks](https://github.com/indutny/bn.js/issues/89)
  3. Used by the Ethereum foundation across all [`ethereumjs`](https://github.com/ethereumjs) repositories
  4. Is already used by a critical JS dependency of many ethereum packages, see package [`elliptic`](https://github.com/indutny/elliptic)
  5. Does not support decimals or floats (for greater precision), remember, the Ethereum blockchain cannot and will not support float values or decimal numbers

## Contributing

Please help better the ecosystem by submitting issues and pull requests to default. We need all the help we can get to build the absolute best linting standards and utilities. We follow the AirBNB linting standard and the unix philosophy.

## Guides

You'll find more detailed information on using `ethjs-unit` and tailoring it to your needs in our guides:

- [User guide](docs/user-guide.md) - Usage, configuration, FAQ and complementary tools.
- [Developer guide](docs/developer-guide.md) - Contributing to `ethjs-unit`, writing coverage and updates.

## Help out

There is always a lot of work to do, and will have many rules to maintain. So please help out in any way that you can:

- Create, enhance, and debug ethjs rules (see our guide to ["Working on rules"](./github/CONTRIBUTING.md)).
- Improve documentation.
- Chime in on any open issue or pull request.
- Open new issues about your ideas for making `ethjs-unit` better, and pull requests to show us how your idea works.
- Add new tests to *absolutely anything*.
- Create or contribute to ecosystem tools, like modules for encoding or contracts.
- Spread the word.

Please consult our [Code of Conduct](CODE_OF_CONDUCT.md) docs before helping out.

We communicate via [issues](https://github.com/ethjs/ethjs-unit/issues) and [pull requests](https://github.com/ethjs/ethjs-unit/pulls).

## Important documents

- [Changelog](CHANGELOG.md)
- [Code of Conduct](CODE_OF_CONDUCT.md)
- [License](https://raw.githubusercontent.com/ethjs/ethjs-unit/master/LICENSE)

## Licence

This project is licensed under the MIT license, Copyright (c) 2016 Nick Dodson. For more information see LICENSE.md.

```
The MIT License

Copyright (c) 2016 Nick Dodson. nickdodson.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
```


CHANGELOG.md
# 0.1.8 -- ethjs-unit

1. TypeScript support
2. Refactor

# 0.1.7 -- ethjs-unit

1. Update toWei
2. Add toWeiByDecimals
3. Add fromWeiByDecimals
4. Update tests

# 0.1.5 -- update bn formatting

1. Update bn formatting

# 0.1.3 -- config fixes

1. webpack config updates
2. build config updates

# 0.1.2 -- removed BigNumber's, replaced with all BN.js

1. new approach with BN.js and absolute precision
2. more coverage
3. more docs

# 0.0.2 -- ethjs-unit

1. More test coverage
2. Config
3. More docs

# 0.0.1 -- ethjs-unit

1. Basic testing
2. Basic docs
3. License
4. Linting
5. Basic exports


tests/index.js
const units = require('../dist/index.js'); // eslint-disable-line
const BigNumber = require('bn.js'); // eslint-disable-line
const ActualBigNumber = require('bignumber.js');
const Web3 = require('web3'); // eslint-disable-line
const web3 = new Web3(); // eslint-disable-line
const assert = require('chai').assert; // eslint-disable-line
const totalTypes = Object.keys(units.unitMap).length;

function testRandomValueAgainstWeb3ToWei(negative) {
  const stringTestValue = `${negative ? '-' : ''}${String(Math.floor((Math.random() * 100000000000000000) + 1))}`;
  const randomunitsType = Object.keys(units.unitMap)[Math.floor((Math.random() * (totalTypes - 1)) + 1)];
  const unitsValue = units.toWei(stringTestValue, randomunitsType);
  const web3Value = web3.toWei(stringTestValue, randomunitsType);

  // it(`toWei should work like web3 val ${unitsValue} should equal ${web3Value}`, () => {
  assert.deepEqual(unitsValue, web3Value);
  // });
}

function testRandomValueAgainstWeb3FromWei(negative) {
  const stringTestValue = `${negative ? '-' : ''}${String(Math.floor((Math.random() * 100000000000000000) + 1))}`;
  const randomunitsType = Object.keys(units.unitMap)[Math.floor((Math.random() * (totalTypes - 1)) + 1)];
  const unitsValue = units.fromWei(stringTestValue, randomunitsType);
  const web3Value = web3.fromWei(stringTestValue, randomunitsType);

  // it(`fromWei should work like web3 rounded val ${unitsValue.substr(0, web3Value.length - 1)} should equal ${web3Value.substr(0, web3Value.length - 1)} for unit type ${randomunitsType}`, () => {
  assert.deepEqual(unitsValue.substr(0, web3Value.length - 1), web3Value.substr(0, web3Value.length - 1));
  // });
}

describe('getValueOfUnit', () => {
  it('should throw when undefined or not string', () => {
    function invalidFromWei() {
      units.fromWei(1000000000000000000, 'something');
    }
    assert.throws(invalidFromWei, Error);
  });
});

describe('toWei', () => {
  it('should handle edge cases', () => {
    assert.equal(units.toWei(0, 'wei'), '0');
    assert.equal(units.toWei('0.0', 'wei'), '0');
    assert.equal(units.toWei('.3', 'ether'), '300000000000000000');
    assert.throws(() => units.toWei('.', 'wei'), Error);
    assert.throws(() => units.toWei('1.243842387924387924897423897423', 'ether'), Error);
    assert.throws(() => units.toWei('8723.98234.98234', 'ether'), Error);
  });

  it('should return the correct value', () => {
    assert.equal(units.toWei(1, 'wei'), '1');
    assert.equal(units.toWei(1, 'kwei'), '1000');
    assert.equal(units.toWei(1, 'Kwei'), '1000');
    assert.equal(units.toWei(1, 'babbage'), '1000');
    assert.equal(units.toWei(1, 'mwei'), '1000000');
    assert.equal(units.toWei(1, 'Mwei'), '1000000');
    assert.equal(units.toWei(1, 'lovelace'), '1000000');
    assert.equal(units.toWei(1, 'gwei'), '1000000000');
    assert.equal(units.toWei(1, 'Gwei'), '1000000000');
    assert.equal(units.toWei(1, 'shannon'), '1000000000');
    assert.equal(units.toWei(1, 'szabo'), '1000000000000');
    assert.equal(units.toWei(1, 'finney'), '1000000000000000');
    assert.equal(units.toWei(1, 'ether'), '1000000000000000000');
    assert.equal(units.toWei(1, 'kether'), '1000000000000000000000');
    assert.equal(units.toWei(1, 'grand'), '1000000000000000000000');
    assert.equal(units.toWei(1, 'mether'), '1000000000000000000000000');
    assert.equal(units.toWei(1, 'gether'), '1000000000000000000000000000');
    assert.equal(units.toWei(1, 'tether'), '1000000000000000000000000000000');

    assert.equal(units.toWei(1, 'kwei'), units.toWei(1, 'femtoether'));
    assert.equal(units.toWei(1, 'szabo'), units.toWei(1, 'microether'));
    assert.equal(units.toWei(1, 'finney'), units.toWei(1, 'milliether'));
    assert.equal(units.toWei(1, 'milli'), units.toWei(1, 'milliether'));
    assert.equal(units.toWei(1, 'milli'), units.toWei(1000, 'micro'));

    assert.throws(() => { units.toWei(1, 'wei1'); }, Error);
  });
});

describe('fromWei', () => {
  it('should handle options', () => {
    assert.equal(units.fromWei(10000000, 'wei', { commify: true }), '10,000,000');
  });

  it('should return the correct value', () => {
    assert.equal(units.fromWei(1000000000000000000, 'wei'), '1000000000000000000');
    assert.equal(units.fromWei(1000000000000000000, 'kwei'), '1000000000000000');
    assert.equal(units.fromWei(1000000000000000000, 'mwei'), '1000000000000');
    assert.equal(units.fromWei(1000000000000000000, 'gwei'), '1000000000');
    assert.equal(units.fromWei(1000000000000000000, 'szabo'), '1000000');
    assert.equal(units.fromWei(1000000000000000000, 'finney'), '1000');
    assert.equal(units.fromWei(1000000000000000000, 'ether'), '1');
    assert.equal(units.fromWei(1000000000000000000, 'kether'), '0.001');
    assert.equal(units.fromWei(1000000000000000000, 'grand'), '0.001');
    assert.equal(units.fromWei(1000000000000000000, 'mether'), '0.000001');
    assert.equal(units.fromWei(1000000000000000000, 'gether'), '0.000000001');
    assert.equal(units.fromWei(1000000000000000000, 'tether'), '0.000000000001');
  });
});

describe('units', () => {
  describe('normal functionality', () => {
    it('should be the same as web3', () => {
      for (var i = 0; i < 15000; i++) { // eslint-disable-line
        testRandomValueAgainstWeb3ToWei(false);
        testRandomValueAgainstWeb3ToWei(true);
        testRandomValueAgainstWeb3FromWei(false);
        testRandomValueAgainstWeb3FromWei(true);
      }
    });
  });
});

src/tests/test.index.js
const units = require('../index.js'); // eslint-disable-line
const BigNumber = require('bn.js'); // eslint-disable-line
const ActualBigNumber = require('bignumber.js');
const Web3 = require('web3'); // eslint-disable-line
const web3 = new Web3(); // eslint-disable-line
const assert = require('chai').assert; // eslint-disable-line
const totalTypes = Object.keys(units.unitMap).length;

function testRandomValueAgainstWeb3ToWei(negative) {
  const stringTestValue = `${negative ? '-' : ''}${String(Math.floor((Math.random() * 100000000000000000) + 1))}`;
  const randomunitsType = Object.keys(units.unitMap)[Math.floor((Math.random() * (totalTypes - 1)) + 1)];
  const unitsValue = units.toWei(stringTestValue, randomunitsType);
  const web3Value = web3.toWei(stringTestValue, randomunitsType);

  // it(`toWei should work like web3 val ${unitsValue} should equal ${web3Value}`, () => {
  assert.deepEqual(unitsValue, web3Value);
  // });
}

function testRandomValueAgainstWeb3FromWei(negative) {
  const stringTestValue = `${negative ? '-' : ''}${String(Math.floor((Math.random() * 100000000000000000) + 1))}`;
  const randomunitsType = Object.keys(units.unitMap)[Math.floor((Math.random() * (totalTypes - 1)) + 1)];
  const unitsValue = units.fromWei(stringTestValue, randomunitsType);
  const web3Value = web3.fromWei(stringTestValue, randomunitsType);

  // it(`fromWei should work like web3 rounded val ${unitsValue.substr(0, web3Value.length - 1)} should equal ${web3Value.substr(0, web3Value.length - 1)} for unit type ${randomunitsType}`, () => {
  assert.deepEqual(unitsValue.substr(0, web3Value.length - 1), web3Value.substr(0, web3Value.length - 1));
  // });
}

describe('getValueOfUnit', () => {
  it('should throw when undefined or not string', () => {
    function invalidFromWei() {
      units.fromWei(1000000000000000000, 'something');
    }
    assert.throws(invalidFromWei, Error);
  });
});

describe('toWei', () => {
  it('should handle edge cases', () => {
    assert.equal(units.toWei(0, 'wei'), '0');
    assert.equal(units.toWei('0.0', 'wei'), '0');
    assert.equal(units.toWei('.3', 'ether'), '300000000000000000');
    assert.throws(() => units.toWei('.', 'wei'), Error);
    assert.throws(() => units.toWei('1.243842387924387924897423897423', 'ether'), Error);
    assert.throws(() => units.toWei('8723.98234.98234', 'ether'), Error);
  });

  it('should return the correct value', () => {
    assert.equal(units.toWei(1, 'wei'), '1');
    assert.equal(units.toWei(1, 'kwei'), '1000');
    assert.equal(units.toWei(1, 'Kwei'), '1000');
    assert.equal(units.toWei(1, 'babbage'), '1000');
    assert.equal(units.toWei(1, 'mwei'), '1000000');
    assert.equal(units.toWei(1, 'Mwei'), '1000000');
    assert.equal(units.toWei(1, 'lovelace'), '1000000');
    assert.equal(units.toWei(1, 'gwei'), '1000000000');
    assert.equal(units.toWei(1, 'Gwei'), '1000000000');
    assert.equal(units.toWei(1, 'shannon'), '1000000000');
    assert.equal(units.toWei(1, 'szabo'), '1000000000000');
    assert.equal(units.toWei(1, 'finney'), '1000000000000000');
    assert.equal(units.toWei(1, 'ether'), '1000000000000000000');
    assert.equal(units.toWei(1, 'kether'), '1000000000000000000000');
    assert.equal(units.toWei(1, 'grand'), '1000000000000000000000');
    assert.equal(units.toWei(1, 'mether'), '1000000000000000000000000');
    assert.equal(units.toWei(1, 'gether'), '1000000000000000000000000000');
    assert.equal(units.toWei(1, 'tether'), '1000000000000000000000000000000');

    assert.equal(units.toWei(1, 'kwei'), units.toWei(1, 'femtoether'));
    assert.equal(units.toWei(1, 'szabo'), units.toWei(1, 'microether'));
    assert.equal(units.toWei(1, 'finney'), units.toWei(1, 'milliether'));
    assert.equal(units.toWei(1, 'milli'), units.toWei(1, 'milliether'));
    assert.equal(units.toWei(1, 'milli'), units.toWei(1000, 'micro'));

    assert.throws(() => { units.toWei(1, 'wei1'); }, Error);
  });
});

describe('numberToString', () => {
  it('should handle edge cases', () => {
    // assert.throws(() => units.numberToString(null), Error);
    assert.throws(() => units.numberToString(undefined), Error);
    // assert.throws(() => units.numberToString(NaN), Error);
    assert.throws(() => units.numberToString({}), Error);
    assert.throws(() => units.numberToString([]), Error);
    assert.throws(() => units.numberToString('-1sdffsdsdf'), Error);
    assert.throws(() => units.numberToString('-0..-...9'), Error);
    assert.throws(() => units.numberToString('fds'), Error);
    assert.throws(() => units.numberToString(''), Error);
    assert.throws(() => units.numberToString('#'), Error);
    assert.equal(units.numberToString(55), '55');
    assert.equal(units.numberToString(1), '1');
    assert.equal(units.numberToString(-1), '-1');
    assert.equal(units.numberToString(0), '0');
    assert.equal(units.numberToString(-0), '0');
    assert.equal(units.numberToString(new ActualBigNumber(10.1)), '10.1');
    assert.equal(units.numberToString(new ActualBigNumber(10000)), '10000');
    assert.equal(units.numberToString(new BigNumber(10000)), '10000');
    assert.equal(units.numberToString(new BigNumber('-1')), '-1');
    assert.equal(units.numberToString(new BigNumber('1')), '1');
    assert.equal(units.numberToString(new BigNumber(0)), '0');
  });
});

describe('fromWei', () => {
  it('should handle options', () => {
    assert.equal(units.fromWei(10000000, 'wei', { commify: true }), '10,000,000');
  });

  it('should return the correct value', () => {
    assert.equal(units.fromWei(1000000000000000000, 'wei'), '1000000000000000000');
    assert.equal(units.fromWei(1000000000000000000, 'kwei'), '1000000000000000');
    assert.equal(units.fromWei(1000000000000000000, 'mwei'), '1000000000000');
    assert.equal(units.fromWei(1000000000000000000, 'gwei'), '1000000000');
    assert.equal(units.fromWei(1000000000000000000, 'szabo'), '1000000');
    assert.equal(units.fromWei(1000000000000000000, 'finney'), '1000');
    assert.equal(units.fromWei(1000000000000000000, 'ether'), '1');
    assert.equal(units.fromWei(1000000000000000000, 'kether'), '0.001');
    assert.equal(units.fromWei(1000000000000000000, 'grand'), '0.001');
    assert.equal(units.fromWei(1000000000000000000, 'mether'), '0.000001');
    assert.equal(units.fromWei(1000000000000000000, 'gether'), '0.000000001');
    assert.equal(units.fromWei(1000000000000000000, 'tether'), '0.000000000001');
  });
});

describe('units', () => {
  describe('normal functionality', () => {
    it('should be the same as web3', () => {
      for (var i = 0; i < 15000; i++) { // eslint-disable-line
        testRandomValueAgainstWeb3ToWei(false);
        testRandomValueAgainstWeb3ToWei(true);
        testRandomValueAgainstWeb3FromWei(false);
        testRandomValueAgainstWeb3FromWei(true);
      }
    });
  });
});


src/index.ts
/**
	Primary Attribution
	Richard Moore <ricmoo@me.com>
	https://github.com/ethers-io

	Fork and Refactor by Normalizex
	https://github.com/Normalizex
*/

import BN from 'bn.js';

const unitMap = {
	noether:      0,
	wei:          0,
	kwei:         3, 
	Kwei:         3, 
	babbage:      3, 
	femtoether:   3, 
	mwei:         6, 
	Mwei:         6, 
	lovelace:     6, 
	picoether:    6, 
	gwei:         9, 
	Gwei:         9, 
	shannon:      9, 
	nanoether:    9, 
	nano:         9, 
	szabo:        12, 
	microether:   12, 
	micro:        12, 
	finney:       15, 
	milliether:   15, 
	milli:        15, 
	ether:        18, 
	kether:       21, 
	grand:        21, 
	mether:       24, 
	gether:       27, 
	tether:       30, 
};

/**
 * Returns value of unit in Wei
 *
 * @param unitInput -  the unit to convert to, default ether
 * @returns value of the unit (in Wei)
 * @throws error if the unit is not correct:w
*/
const getDecimalsOfUnit = (unitInput: string | number) => {
	const unit = unitInput.toString();

	const unitValue = Object
		.entries(unitMap)
		.find((content) =>
			content[0] === unit ||
			content[1] === Number(unit)
		)?.slice(-1)[0];

	if (typeof unitValue === 'undefined')
		throw new Error(`[ethjs-unit] the unit provided ${unitInput} doesn't exists, please use the one of the following units ${JSON.stringify(unitMap, null, 2)}`);
  
	return Number(unitValue);
}

/**
 * @param numberInput
 * @param decimals - Number of zeros
 * @returns number in units
*/
export const toWeiByDecimals = (numberInput: string | number, decimals: number | string) => {
	if (isNaN(Number(decimals)))
		throw new Error(`[ethjs-unit] while converting number ${numberInput} to wei, invalid decimals`);

	const unitValue = `1${'0'.repeat(Number(decimals))}`;
	const base = new BN(unitValue, 10);
	const baseLength = unitValue.length - 1 || 1;

	let etherValue = numberInput.toString().slice();

	const negative = (etherValue.substring(0, 1) === '-');
	if (negative) etherValue = etherValue.substring(1);
	if (etherValue === '.')
		throw new Error(`[ethjs-unit] while converting number ${numberInput} to wei, invalid value`);

	const comps = etherValue.split('.');
	if (comps.length > 2)
		throw new Error(`[ethjs-unit] while converting number ${numberInput} to wei,  too many decimal points`);

	const [whole = '0', fraction = '0'] = comps;

	if (fraction.length > baseLength)
		throw new Error(`[ethjs-unit] while converting number ${numberInput} to wei, too many decimal places`);
	
	const num = new BN(whole)
		.mul(base)
		.add(new BN(fraction + '0'.repeat(baseLength - fraction.length)));

	return (negative ? num.mul(new BN(-1)) : num).toString(10);
}

interface fromWeiOptions {
	pad?: boolean,
	commify?: boolean
}

export const fromWeiByDecimals = (
	numberInput: string | number,
	decimals: number | string,
	options: fromWeiOptions = {}
) => {
	if (isNaN(Number(decimals)))
		throw new Error(`[ethjs-unit] while converting number ${numberInput} to wei, invalid decimals`);

	const unitValue = `1${'0'.repeat(Number(decimals))}`;
	const base = new BN(unitValue, 10);
	const baseLength = unitValue.length - 1 || 1;
	const negative = new BN(numberInput.toString()).lt(new BN(0));
	const wei = negative
		? new BN(numberInput.toString()).mul(new BN(-1))
		: new BN(numberInput.toString())

	let fraction = wei.mod(base).toString(10);
	let whole = wei.div(base).toString(10);

	while (fraction.length < baseLength) {
		fraction = `0${fraction}`;
	}

	if (!options.pad) fraction = fraction.match(/^([0-9]*[1-9]|0)(0*)/)?.slice(0)[1] || fraction;
	if (options.commify) whole = whole.replace(/\B(?=(\d{3})+(?!\d))/g, ',');
	
	return `${negative ? "-": ""}${whole}${fraction == '0' ? '' : `.${fraction}`}`;
};

/**
 * @param numberInput 
 * @param unit - The name or value of unitMap
 * - 'ether', 'wei'
 * - '18', 21, '3'
 * @returns number in units
*/
export const toWei = (numberInput: string | number, unit: string) =>
	toWeiByDecimals(numberInput, getDecimalsOfUnit(unit));

/**
 * @param unitInput 
 * @param unit - The name or value of unitMap
 * - 'ether', 'wei'
 * - '18', 21, '3'
 * @returns number in units
*/
export const fromWei = (
	unitInput: string | number,
	unit: string,
	options: fromWeiOptions = {}
) =>
	fromWeiByDecimals(unitInput, getDecimalsOfUnit(unit), options);

export {
	unitMap,
	getDecimalsOfUnit,
}

src/index.js
/*
Primary Attribution
Richard Moore <ricmoo@me.com>
https://github.com/ethers-io

Note, Richard is a god of ether gods. Follow and respect him, and use Ethers.io!
*/

const BN = require('bn.js');
const numberToBN = require('number-to-bn');

const zero = new BN(0);
const negative1 = new BN(-1);

// complete ethereum unit map
const unitMap = {
  'noether':      '0', // eslint-disable-line
  'wei':          '1', // eslint-disable-line
  'kwei':         '1000', // eslint-disable-line
  'Kwei':         '1000', // eslint-disable-line
  'babbage':      '1000', // eslint-disable-line
  'femtoether':   '1000', // eslint-disable-line
  'mwei':         '1000000', // eslint-disable-line
  'Mwei':         '1000000', // eslint-disable-line
  'lovelace':     '1000000', // eslint-disable-line
  'picoether':    '1000000', // eslint-disable-line
  'gwei':         '1000000000', // eslint-disable-line
  'Gwei':         '1000000000', // eslint-disable-line
  'shannon':      '1000000000', // eslint-disable-line
  'nanoether':    '1000000000', // eslint-disable-line
  'nano':         '1000000000', // eslint-disable-line
  'szabo':        '1000000000000', // eslint-disable-line
  'microether':   '1000000000000', // eslint-disable-line
  'micro':        '1000000000000', // eslint-disable-line
  'finney':       '1000000000000000', // eslint-disable-line
  'milliether':   '1000000000000000', // eslint-disable-line
  'milli':        '1000000000000000', // eslint-disable-line
  'ether':        '1000000000000000000', // eslint-disable-line
  'kether':       '1000000000000000000000', // eslint-disable-line
  'grand':        '1000000000000000000000', // eslint-disable-line
  'mether':       '1000000000000000000000000', // eslint-disable-line
  'gether':       '1000000000000000000000000000', // eslint-disable-line
  'tether':       '1000000000000000000000000000000', // eslint-disable-line
};

/**
 * Returns value of unit in Wei
 *
 * @method getValueOfUnit
 * @param {String} unit the unit to convert to, default ether
 * @returns {BigNumber} value of the unit (in Wei)
 * @throws error if the unit is not correct:w
 */
function getValueOfUnit(unitInput) {
  const unit = unitInput ? unitInput.toLowerCase() : 'ether';
  var unitValue = unitMap[unit]; // eslint-disable-line

  if (typeof unitValue !== 'string') {
    throw new Error(`[ethjs-unit] the unit provided ${unitInput} doesn't exists, please use the one of the following units ${JSON.stringify(unitMap, null, 2)}`);
  }

  return new BN(unitValue, 10);
}

function numberToString(arg) {
  if (typeof arg === 'string') {
    if (!arg.match(/^-?[0-9.]+$/)) {
      throw new Error(`while converting number to string, invalid number value '${arg}', should be a number matching (^-?[0-9.]+).`);
    }
    return arg;
  } else if (typeof arg === 'number') {
    return String(arg);
  } else if (typeof arg === 'object' && arg.toString && (arg.toTwos || arg.dividedToIntegerBy)) {
    if (arg.toPrecision) {
      return String(arg.toPrecision());
    } else { // eslint-disable-line
      return arg.toString(10);
    }
  }
  throw new Error(`while converting number to string, invalid number value '${arg}' type ${typeof arg}.`);
}

function fromWei(weiInput, unit, optionsInput) {
  var wei = numberToBN(weiInput); // eslint-disable-line
  var negative = wei.lt(zero); // eslint-disable-line
  const base = getValueOfUnit(unit);
  const baseLength = unitMap[unit].length - 1 || 1;
  const options = optionsInput || {};

  if (negative) {
    wei = wei.mul(negative1);
  }

  var fraction = wei.mod(base).toString(10); // eslint-disable-line

  while (fraction.length < baseLength) {
    fraction = `0${fraction}`;
  }

  if (!options.pad) {
    fraction = fraction.match(/^([0-9]*[1-9]|0)(0*)/)[1];
  }

  var whole = wei.div(base).toString(10); // eslint-disable-line

  if (options.commify) {
    whole = whole.replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  }

  var value = `${whole}${fraction == '0' ? '' : `.${fraction}`}`; // eslint-disable-line

  if (negative) {
    value = `-${value}`;
  }

  return value;
}

function fromWeiByDecimals(weiInput, unitDecimals, optionsInput) {
  var wei = numberToBN(weiInput); // eslint-disable-line
  var negative = wei.lt(zero); // eslint-disable-line
  const unitValue = `1${'0'.repeat(Number(unitDecimals))}`;
  const base = new BN(unitValue, 10);
  const baseLength = unitValue.length - 1 || 1;
  const options = optionsInput || {};

  if (negative) {
    wei = wei.mul(negative1);
  }

  var fraction = wei.mod(base).toString(10); // eslint-disable-line

  while (fraction.length < baseLength) {
    fraction = `0${fraction}`;
  }

  if (!options.pad) {
    fraction = fraction.match(/^([0-9]*[1-9]|0)(0*)/)[1];
  }

  var whole = wei.div(base).toString(10); // eslint-disable-line

  if (options.commify) {
    whole = whole.replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  }

  var value = `${whole}${fraction == '0' ? '' : `.${fraction}`}`; // eslint-disable-line

  if (negative) {
    value = `-${value}`;
  }

  return value;
}

function toWei(etherInput, unit) {
  var ether = numberToString(etherInput); // eslint-disable-line
  const base = getValueOfUnit(unit);
  const baseLength = unitMap[unit].length - 1 || 1;

  // Is it negative?
  var negative = (ether.substring(0, 1) === '-'); // eslint-disable-line
  if (negative) {
    ether = ether.substring(1);
  }

  if (ether === '.') { throw new Error(`[ethjs-unit] while converting number ${etherInput} to wei, invalid value`); }

  // Split it into a whole and fractional part
  var comps = ether.split('.'); // eslint-disable-line
  if (comps.length > 2) { throw new Error(`[ethjs-unit] while converting number ${etherInput} to wei,  too many decimal points`); }

  var whole = comps[0], fraction = comps[1]; // eslint-disable-line

  if (!whole) { whole = '0'; }
  if (!fraction) { fraction = '0'; }
  if (fraction.length > baseLength) { throw new Error(`[ethjs-unit] while converting number ${etherInput} to wei, too many decimal places`); }

  while (fraction.length < baseLength) {
    fraction += '0';
  }

  whole = new BN(whole);
  fraction = new BN(fraction);
  var wei = (whole.mul(base)).add(fraction); // eslint-disable-line

  if (negative) {
    wei = wei.mul(negative1);
  }

  return new BN(wei.toString(10), 10).toString(10);
}

function toWeiByDecimals(etherInput, unitDecimals) {
  var ether = numberToString(etherInput); // eslint-disable-line
  const unitValue = `1${'0'.repeat(Number(unitDecimals))}`;
  const base = new BN(unitValue, 10);
  const baseLength = unitValue.length - 1 || 1;

  // Is it negative?
  var negative = (ether.substring(0, 1) === '-'); // eslint-disable-line
  if (negative) {
    ether = ether.substring(1);
  }

  if (ether === '.') { throw new Error(`[ethjs-unit] while converting number ${etherInput} to wei, invalid value`); }

  // Split it into a whole and fractional part
  var comps = ether.split('.'); // eslint-disable-line
  if (comps.length > 2) { throw new Error(`[ethjs-unit] while converting number ${etherInput} to wei,  too many decimal points`); }

  var whole = comps[0], fraction = comps[1]; // eslint-disable-line

  if (!whole) { whole = '0'; }
  if (!fraction) { fraction = '0'; }
  if (fraction.length > baseLength) { throw new Error(`[ethjs-unit] while converting number ${etherInput} to wei, too many decimal places`); }

  while (fraction.length < baseLength) {
    fraction += '0';
  }

  whole = new BN(whole);
  fraction = new BN(fraction);
  var wei = (whole.mul(base)).add(fraction); // eslint-disable-line

  if (negative) {
    wei = wei.mul(negative1);
  }

  return new BN(wei.toString(10), 10).toString(10);
}

module.exports = {
  unitMap,
  numberToString,
  getValueOfUnit,
  fromWei,
  toWei,
  fromWeiByDecimals,
  toWeiByDecimals,
};
